src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java

```diff
@@ -22,23 +24,36 @@ public void apply(Project project) {
         if (System.getenv(TEAMCITY_VERSION_ENV) != null && System.getenv(GRADLE_BUILDSCAN_TEAMCITY_PLUGIN_ENV) == null) {
             project.getLogger().quiet(ServiceMessage.of(BUILD_SCAN_SERVICE_MESSAGE_NAME, BUILD_SCAN_SERVICE_STARTED_MESSAGE_ARGUMENT).toString());
 
-            BuildScanExtension buildScanExtension = project.getExtensions().findByType(BuildScanExtension.class);
-            if (supportsScanPublishedListener(buildScanExtension)) {
-                buildScanExtension.buildScanPublished(publishedBuildScan -> {
-                            ServiceMessage serviceMessage = ServiceMessage.of(
-                                    BUILD_SCAN_SERVICE_MESSAGE_NAME,
-                                    BUILD_SCAN_SERVICE_URL_MESSAGE_ARGUMENT_PREFIX + publishedBuildScan.getBuildScanUri().toString()
-                            );
-                            project.getLogger().quiet(serviceMessage.toString());
-                        }
-                );
+            ExtensionContainer extensions = project.getExtensions();
+            try {
+                Class<?> buildScanExtensionClass = Class.forName(BUILD_SCAN_EXTENSION);
+                if (extensions.findByType(buildScanExtensionClass) != null) {
+                    registerCallback(project);
+                    return;
+                }
+            } catch (ClassNotFoundException ignore) {
             }
+            project.getPluginManager().withPlugin(BUILD_SCAN_PLUGIN_ID, appliedPlugin -> registerCallback(project));
         }
     }
 
-    private static boolean supportsScanPublishedListener(@Nullable BuildScanExtension extension) {
-        if (extension == null) return false;
-        for (Method method : extension.getClass().getMethods()) {
+    private static void registerCallback(Project project) {
+        BuildScanExtension buildScanExtension = project.getExtensions().getByType(BuildScanExtension.class);
```

Comment:
We could get to this line even with `Class.forName(BUILD_SCAN_EXTENSION)` having thrown an exception.

Reply:
reworked, but this was not a problem